cnap/core/streambroker.py

```diff
@@ -0,0 +1,209 @@
+                sleep_time *= 2
+
+                time.sleep(sleep_time)
+
+                # Connect to Redis server.
+                self._conn = redis.Redis(host=host, port=port, db=0)
+
+                # Connection is not successful, can not return.
+                continue
+
+            # If no connection error occurs, return directly.
+            return
+
+    def publish_frame(self, topic: str, frame: Frame) -> None:
+        if topic is None:
+            raise ValueError("topic can not be None")
+        if frame is None:
+            raise ValueError("frame can not be None")
+        try:
+            _, img = cv2.imencode('.jpg', frame.raw)
+        except Exception as e:
+            raise RuntimeError(f"Error during encoding image into jpg format: {str(e)}") from e
+        frame.raw = img
+        try:
+            frame_blob = frame.to_blob()
+        except RuntimeError as e:
+            raise RuntimeError(e) from e
+        self._conn.publish(topic, frame_blob)
+
+
+class KafkaStreamBrokerClient(StreamBrokerClientBase):
+
+    """
+    Kafka implementation for stream broker client.
+    """
+
+    def __init__(self):
+        self._conn = None
+
+    def connect(self, host="127.0.0.1", port=9092):
+        """
+        Connect to kafka server, will attempt to reconnect when a NoBrokersAvailable error occcurs.
+
+        Args:
+            host: The host hostname/ip of Kafka server.
+            port: The port of Kafka server.
+
+        Returns: None
+
+        Raises:
+            kafka.errors.NoBrokersAvailable: if connection to the Kafka server fails and
+                reconnection exceeds the limit.
+        """
+        address = host + ":" + str(port)
+
+        sleep_time = 0.5
+
+        # Attempts to reconnect when a NoBrokersAvailable error occurs, up to MAX_RECONNECTION_TIMES
+        # times, so if a NoBrokersAvailable error still occurs on the (MAX_RECONNECTION_TIMES + 1)th
+        # loop, raise NoBrokersAvailable error
+        for i in range(self.MAX_RECONNECTION_TIMES + 1):
+            try:
+                self._conn = KafkaProducer(bootstrap_servers=address)
+            except kafka.errors.NoBrokersAvailable as e:
+                if i == self.MAX_RECONNECTION_TIMES:
+                    LOG.error("Continuously reconnect to Kafka server more than %d times",
+                              self.MAX_RECONNECTION_TIMES)
+                    raise kafka.errors.NoBrokersAvailable(e) from e
+
+                LOG.error("Failed to connect to Kafka: %s", e)
+                LOG.info("Reconnect to Kafka server")
+
+                # Sleep before reconnect, double the sleep time per reconnection.
+                sleep_time *= 2
+
+                time.sleep(sleep_time)
+
+                # Continue to reconnect to Kafka server
+                continue
+
+            # If no NoBrokersAvailable occurs, return directly.
+            return
```

Comment:
Is it necessary?

Reply:
It is necessary, if connection is successfully, return directly, instead of rejoining a loop. We can also use `break`, but I think `return` is more clear.